### ISSUE TITLE
New Changelog formatting

### ISSUE THREAD

**Issue by blyxyas** [rust-lang/rust-clippy#10847]
\[Related to this [Zulip Meeting](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Meeting.202023-05-30)\]

## Grouping issues

This issue is a discussion around changing the changelog's formatting. Several solutions came out in the 2023-05-30 meeting, this issue discusses the last one (the conclusive one).

The conclusion that we reached on this meeting around the changelog is that we should group all small fixes (usually fixing false-positives and very small suggestion changes), that the user will have to click on. This can be done using the `<details>` and `<summary>` tags.

(The person making the changelog will decide what is a small change)

We would keep the current sections, and add a new one.
(Note that it is tempting using the `Others` section to group all small commits, but sometimes, changes that cannot be categorized in any of the other categories are significant. So we'll have to create a new section)

```
### New Lints
### Moves and Deprecations
### Enhancements
### False Positive Fixes
### Suggestion Fixes/Improvements
### ICE Fixes
### Documentation Improvements
### Others

<details><summary><h3>Minor fixes</h3></summary>
</details>
```

<details><summary><h3>Minor fixes</h3></summary>
    
- Someone fixed a typo yay :D #100000

</details>

Note the double space between `### Others` and `<details>`. And the space between `<details>`, the content, and the final `</details>` (Github is pretty picky when rendering Markdown files that use HTML, but there isn't any other way to create dropdowns in Markdown)

---

## The PR range

We would also replace the commit range (which is not very significant, as a single PR can have many commits) with a new merged PRs range.

The filter to use here would be something like this:

```
is:pr is:closed merged:<last release, YYYY-MM-DD>..<this release> base:master sort:merged-desc 
```

We could display it raw as a range:
#10669..#10841 

(This approach doesn't make much sense, as PRs aren't sequential in any way)

Or maybe just provide a link to view it in Github

[**View 80 PRs merged since 1.69**](https://github.com/rust-lang/rust-clippy/pulls?q=is%3Apr+is%3Aclosed+merged%3A2023-04-20..2023-06-01+base%3Amaster+sort%3Amerged-desc+)

---

Joining these 2 concepts would give us a much more legible changelog.

cc @xFrednet

**Comment by xFrednet** [comment]
Hey @blyxyas, thank you for creating the issue. I really like the PR range and will switch to it in the next changelog.

---

For the second part, about minor changes, you wrote:

> The conclusion that we reached on this meeting around the changelog is that we should group all small fixes (usually fixing false-positives and very small suggestion changes), that the user will have to click on. This can be done using the `<details>` and `<summary>` tags.

I had a different understanding, that minor changes can be dropped completely from the changelog. Having a condensed version was suggested, but there was no real answer, if many people would be looking in into the minor changes. (I usually don't). I guess we'll see if someone requests more details on the next changelog.

Now, I'm also biased to having a shorter changelog format, than just changing it. This has two reasons:
1. Most of the time needed to write the changelog is spent on bringing all PR entries into a (kind of) uniform style. Dropping smaller changes would save time, but if we include them in a condensed section, they still need to be reviewed and often reformulated.
    
    In this regard, it might be worth mentioning, that I'm a relatively slow reader and writer. That, makes me biased towards having to write less ^^
2. The changelog file is quite large by now. My fan laptop fan turns on when I just open the file ^^. By reducing the format length, I'm hoping to slow down the growth. This is of course not the main reason, and there are other options to deal with it. However, slowing down the growth would be a nice benefit.

**Comment by xFrednet** [comment]
Naw, I'll keep em :D

We can see, if we get any feedback from others, when we try the new format in the next changelog. That'll be next release. I'll assign myself to this issue in the meantime, to find it again later. Thank you for documenting this!!!

@rustbot claim

**Comment by blyxyas** [comment]
Sooo... You just pulled an Uno reverse card just to work more?
What a chad :clap:

**Comment by xFrednet** [comment]
What? o.O

Ohhh, you were suggesting updating the old changelogs as well? Sorry, I totally misunderstood the first sentence :sweat_smile:. But if we mainly remove sections, it should be doable with a script right? xD

**Comment by blyxyas** [comment]
@xFrednet Should old commit ranges be converted (possibly by a script) to the new PR ranges? I find them too useful not to use.

**Comment by xFrednet** [comment]
Hey @blyxyas, You can convert the old rages if you like. :D One thing, worth mentioning, is that the commit ranges and dates don't fully align with the releases. Instead, they're determined by the time a sync was done into the rust repo. It should (hopefully) be sufficient to just look up the date of these commits and use that for the date filter :)

**Comment by Alexendoo** [comment]
The new ranges are incorrect, e.g. if you click on the one under 1.70 none of the PRs that I checked are in 1.70. The same goes for the other releases

**Comment by blyxyas** [comment]
Weird, maybe an off-by-one error? How did you check in which versions the PRs were included?

**Comment by Alexendoo** [comment]
I looked at when they were added to the changelog for e.g. new lint PRs

To correctly handle it we would probably want to add it to `util/fetch_prs_between.sh`, getting the merge datetime (not just the date) from github of the first and last merges, that could be

```sh
gh pr view xxxx --json mergedAt -q .mergedAt
```

Or some manual curling maybe if we don't want to depend on `gh`

**Comment by blyxyas** [comment]
Should a cat video be assigned with each release. It would be an absolute win, imo.
1.72's GIF should be this one.

![cat gif](https://cataas.com/cat/SbbeZwoC81vSTzBX)

**Comment by blyxyas** [comment]
I'm getting my cat gifs ready for consumption with the `cataas.com` (Cats As A Service) web, then getting the `src` attribute in the `<img>` element with some Inspect Element. You can do the same with Tenor gifs.

![cat](https://media.tenor.com/4qOlzc3baYgAAAAC/ezzies-cult.gif)

Also, yes I was thinking about adding it to the file (a single line of Markdown), but in a few releases scrolling it would take a lot. Maybe this idea could be exported to the releases team or just let the user scroll through the user as normal?

**Comment by xFrednet** [comment]
Ahh, that explains it. Mine was just a copied link from discord. I have the feeling that opening the changelog file already legs a bit, that's why I'm a bit hesitant with adding it directly to the changelog file.

I don't believe that the release time would accept it, as the changelog blog posts are usually written in a serious and informative manner, but asking wouldn't cost anything, I guess xD

**Comment by blyxyas** [comment]
Maybe we can implement this in a few changelog PRs before proposing it to the release team, so that they see that this is **serious business**.

**Comment by xFrednet** [comment]
Well, have you seen the last changelog PR? ^^ https://github.com/rust-lang/rust-clippy/pull/11351

**Comment by blyxyas** [comment]
I'll have to write an RFC and present it by Christmas (everyone will be charmed by the christmas spirit) so they'll have no option but to accept.

We'll have to make some graphics describing with this change is needed, even more, why it's **vital** to the Rust ecosystem. People now can refer to their favourite release with "that cute businesscat with a tie" instead of the boring old numbers "1.79.0". This is simply the future. This is how things should be.

**Comment by blyxyas** [comment]
In fact, maybe we should start some communication with some Cat Api (maybe Cat As A Service, maybe The Cat Api) so they let us use their API for free (+ premium cat pics :money_mouth_face: )

**Comment by blyxyas** [comment]
Ok, I wrote the draft for this RFC 3116
https://hackmd.io/@G9rSdsSGS3K463E3hsfMYw/Hkd3tzfTh
